### DISCUSSION TITLE
signing key rotation

### DISCUSSION THREAD

**Discussion by avdhoot** [postalserver/postal#1969] [Help with using Postal]
By default postal create 1024 bit key signing key. Now we need 2048 bit key. Is there a way we can rotate this key without breaking postal server? 

Our plan is:
1. Create the new key with 2048 
2. replace with the existing key(cp). 
3. restart postal server 

The problem we see:
1.  Will this update the domain DKIM record on the domain setup page. So can update DNS.
2. any other problem you see?

**Comment by willpower232** [comment]
You can see that individual domains keys are generated and stored by Postal so if you want to change that to 2048 that is a separate thing entirely and you'd also have to be sure that the larger key could be stored within the database, I'm not sure on the size difference off the top of my head

https://github.com/postalserver/postal/blob/9f0697f194209f5fae5e451ba8fb888413fe37fa/app/models/domain.rb#L92-L98

I'm not sure what the signing.key file is used for but seems to be picked up here

https://github.com/postalserver/postal/blob/9f0697f194209f5fae5e451ba8fb888413fe37fa/lib/postal/config.rb#L160-L166

**Comment by avdhoot** [comment]
How to create DKIM record 2048 bit

Do not follow the below steps if DKIM records not be updated immediately. after this change, the email server will start using a new private key to sign the email. the recipient email server will not able to verify email because we have not updated public key in DNS records.  

login into the postal server & change user to root

```
$ docker exec -it postal_web_1 bash
$ bin/rails console
$ id = Domain.where(name: 'xyz.com').ids
$ id.length()
# length greater than 1 then we have to find right domain ID. Pass ID into below command
$ d = Domain.find(id[0])
$ d.dkim_private_key = OpenSSL::PKey::RSA.new(2048).to_s
$ d.dkim_record 
$ d.save
```

You can get an updated record from the domain page OR as the output of d.dkim_record
